### DISCUSSION TITLE
How to set and call the functions of pre-defined porosity

### DISCUSSION THREAD

**Discussion by Devin-one** [PMEAL/OpenPNM#2503] [Q&A]
Hello, everyone. I have some questions to ask you. First, how to set the functions of pre-defined porosity , and then how to call the functions of pre-defined porosity  in different locations in the network. Thank you in advance.

**Comment by jgostick** [comment]
There is not pre-defined porosity, so perhaps you need to reword your question. You can calculate porosity by summing up all the pore and throat volumes, the dividing by the domain volume.  This is tricky though because you have to be very careful about how you define the pore volume.  If the pores are treated as spheres then the they will overlap with their neighboring throats so the "lens volume" needs to be taken into account (lens is the are of overlap).  Anyway, my point is that porosity needs to be done carefully, which is why we don't offer a pre-defined function for it.

**Reply by Devin-one** [reply]
Thank you very much for your answer. By the way, if I want to modify the porosity of different positions in the network, how can I achieve my goal? Look forward to your reply again.

**Reply by jgostick** [reply]
Openpnm was designed to allow different domains.  You can do this be creating multiple ``Geometry`` objects.  Then you assign different pore size distribution functions to each object (as demonstrated [here](https://github.com/PMEAL/OpenPNM/blob/v2.8.2/examples/notebooks/geometry/adjusting_pore_size_distributions.ipynb)).  You can also stitch together two different regions with different spacings then stitch them together as shown [here](https://github.com/PMEAL/OpenPNM/blob/v2.8.2/examples/notebooks/networks/manipulation/stitching_and_merging_networks_together.ipynb).

You can also use a network that allows for tighter packing of pores, as illustrated [here](https://github.com/PMEAL/OpenPNM/blob/v2.8.2/examples/notebooks/networks/generation/dual_cubic_lattices.ipynb).  We also have a ``Bravais`` class which is even more flexible, allowing fcc and bcc lattices, but there is not example for this.  You can figure it out via the docstring hopefully.  

Good luck!

**Comment by Devin-one** [comment]
Hello, Professor, in the process of creating geometric objects, can I use several Weibull distributions to achieve different porosity in different network areas. Or there are other ways to achieve my expected goals.

**Reply by jgostick** [reply]
Hi @Liqiwen123.  The point of ``Geometry`` objects is that you can create multiple of them, with each assigned to a different set of pores/throats.  You can then assign different distributions to each one, as you suggest above.  I don't think we have an example on this.  Note that the downside of this is that you then need to create one ``Physics`` object for each ``Geometry``.  This is a huge pain which we have actually removed from V3 (yet to be released).  There is a workaround though...you can create a single physics, and put the ``network`` for the ``geometry`` argument.  Like ``phys = op.physics.Standard(network=pn, phase=air, geometry=pn)``.  This will actually work.

**Reply by jgostick** [reply]
Ahh, I see. The 'geometry' object approach is more about layers and bimodal distributions, but you want more of a gradient effect.  We don't support that directly, but it's do-able.  What you've got above is a map of numbers spanning 1 to 1.5.  I assume this represents the local average pore size value, and that the actual distributions in each region will have some variability?  Anyway, my first suggestion is that you manipulate the 'pore.seed' values then regenerate the pores sizes.  So if you first multiply 'pore.seed' values by 1/1.5 so the max will be 0.666, then you multiply them again by the number field you have above, some will be rescaled back towards 1.0, whil the ones in the blue region will remain 0.666.  This way you'll have large pores, on average, in the yellow region and small in the blue.  The trick is dealing with the spatial locations of pores and the grid locations of your map.  This can be done by taking your 2D map abovea and 'flattening' (``temp = np.flatten(map)``) it, which turns it into a column array.  You should be able just multiply this flattened array with the 'pore.seeds' value, like ``pn['pore.seed'] *= np.flatten(map)``.

**Reply by jgostick** [reply]
The exception you're getting tells you the problem: if you're going to assign pore.seed as a global property on the network, then you must delete if from the Geometry objects.  So just skip the ``geo.add_model(propname='pore.seed,...)`` part and try again?
